lib/dom/xml_entities.nit

```diff
@@ -0,0 +1,586 @@
+			var src_c = src[pos]
+			while src_c != c do
+				pos += 1
+				if pos >= srclen then return -1
+				if src_c == '\n' then
+					line += 1
+					line_start= pos
+				end
+				src_c = src[pos]
+			end
+			var relpos = pos
+			var fnd = true
+			for i in s do
+				if relpos >= srclen then
+					fnd = false
+					break
+				end
+				if src[relpos] != i then
+					pos += 1
+					fnd = false
+					break
+				end
+				relpos += 1
+			end
+			if fnd then return pos
+		end
+	end
+
+	# Parses a full XML document
+	fun parse_document: XMLEntity do
+		var stack = new Array[XMLStartTag]
+		var doc = new XMLDocument
+		loop
+			ignore_whitespaces
+			if pos >= src.length then break
+			if src[pos] == '<' then
+				var tag = read_tag
+				if tag isa XMLStartTag then
+					if stack.is_empty then
+						tag.parent = doc
+					else
+						var st_last = stack.last
+						tag.parent = st_last
+					end
+					stack.push tag
+				else if tag isa XMLEndTag then
+					if stack.is_empty then
+						return new XMLError(location = tag.location, "Missing matching tag for {tag.tag_name}")
+					end
+					var st_last = stack.last
+					if tag.tag_name == st_last.tag_name then
+						st_last.matching = tag
+						tag.matching = st_last
+						stack.pop
+					else
+						var miss = stack.pop
+						return new XMLError("Missing matching tag for {miss.tag_name}", location=miss.location)
+					end
+				else if tag isa XMLError then
+					return tag
+				else
+					if stack.is_empty then
+						tag.parent = doc
+					else
+						tag.parent = stack.last
+					end
+				end
+			else
+				var st = pos
+				var end_pc = ignore_until("<") - 1
+				var pc = new PCDATA(src.substring(st, end_pc - st + 1).trim)
+				if stack.is_empty then
+					pc.parent = doc
+				else
+					pc.parent = stack.last
+				end
+			end
+		end
+		if not stack.is_empty then
+			var miss = stack.pop
+			return new XMLError("Missing matching tag for {miss.tag_name}", location=miss.location)
+		end
+		return doc
+	end
+
+	# Reads the tag starting in `src` at current position
+	private fun read_tag: XMLEntity do
+		var st_loc = new Location(line, line_offset)
+		var c = src[pos]
+		if not c == '<' then return new XMLError(location=st_loc, "Expected start of tag, got {c}")
+		var st = pos
+		pos += 1
+		c = src[pos]
+		if c == '!' then
+			# Special tag
+			return read_special_tag(st_loc)
+		else if c == '?' then
+			# Prolog tag
+			return read_prolog_tag(st_loc)
+		else if c == '/' then
+			# End tag
+			return read_end_tag(st_loc)
+		else
+			# Start tag
+			return read_start_tag(st_loc)
+		end
+	end
+
+	# Reads a Special tag (starting with <!)
+	#
+	# In case of error, returns a `XMLError`
+	private fun read_special_tag(st_loc: Location): XMLEntity do
+		var srclen = src.length
+		pos += 1
+		if (pos + 2) >= srclen then return new XMLError(location=st_loc, "Unexpected EOF on start of Special tag")
+		if src[pos] == '-' and src[pos + 1] == '-' then
+			pos += 2
+			var comst = pos
+			var comlnst = line
```

Comment:
That was supposed to be a `comment line start`, though you're right ! I'll remove it